### PR TITLE
layers: Improve 00124 (image copy overlaps) message

### PR DESCRIPTION
**Pull request by arno-lunarg**
New error message looks like this:

> Validation Error: [ VUID-vkCmdCopyImage-pRegions-00124 ] Object 0: handle = 0x2123bf53290, type = VK_OBJECT_TYPE_COMMAND_BUFFER; Object 1: handle = 0xe7f79a0000000005, type = VK_OBJECT_TYPE_IMAGE; Object 2: handle = 0xe7f79a0000000005, type = VK_OBJECT_TYPE_IMAGE; | MessageID = 0x384abddf | vkCmdCopyImage: pRegion[0] src overlaps with pRegions[0]. Overlap info: { subresource { aspectMask: VK_IMAGE_ASPECT_COLOR_BIT, mipLevel: 0, baseArrayLayer: 0, layerCount: 1 }, offset {0, 0, 0}, extent {64, 64, 1} }. The Vulkan spec states: The union of all source regions, and the union of all destination regions, specified by the elements of pRegions, must not overlap in memory (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkCmdCopyImage-pRegions-00124)

Add test `OverlappingImageCopy`

`VUID-vkCmdCopyImage-pRegions-00124` cannot be checked in the general case as the mapping between an image and its underlying memory is undefined, so closing https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/6325

### CODE REVIEW

File: layers/core_checks/cc_copy_blit_resolve.cpp [hunk 1]
```diff
@@ -18,49 +18,95 @@
...
 #include "generated/vk_enum_string_helper.h"
 #include "generated/chassis.h"
-#include "core_validation.h"
 
-// Returns true if [x, xoffset] and [y, yoffset] overlap
-static bool RangesIntersect(int32_t start, uint32_t start_offset, int32_t end, uint32_t end_offset) {
-    bool result = false;
-    uint32_t intersection_min = std::max(static_cast<uint32_t>(start), static_cast<uint32_t>(end));
-    uint32_t intersection_max = std::min(static_cast<uint32_t>(start) + start_offset, static_cast<uint32_t>(end) + end_offset);
+// Returns true if [x, x + x_offset] and [y, y + y_offset] overlap
```
Comment (artem-lunarg) [inline layers/core_checks/cc_copy_blit_resolve.cpp:29]: this comment is for `RangesIntersect`

File: layers/core_checks/cc_copy_blit_resolve.cpp [hunk 2]
```diff
@@ -18,49 +18,95 @@
...
+    int64_t intersection_max = std::min(x + static_cast<int64_t>(x_offset), y + static_cast<int64_t>(y_offset));
 
-    if (intersection_max > intersection_min) {
-        result = true;
-    }
-    return result;
+    return {intersection_min, intersection_max};
+}
+
+// Returns true if [x, x + x_offset] and [y, y + y_offset] overlap
```
Comment (artem-lunarg) [inline layers/core_checks/cc_copy_blit_resolve.cpp:37]: Typo in the comment, the intervals are semi-closed: `[x, x + x_offset)`, `[y, y + y_offset)`

File: layers/core_checks/cc_copy_blit_resolve.cpp [hunk 3]
```diff
@@ -18,49 +18,95 @@
...
 #include "generated/chassis.h"
-#include "core_validation.h"
 
-// Returns true if [x, xoffset] and [y, yoffset] overlap
-static bool RangesIntersect(int32_t start, uint32_t start_offset, int32_t end, uint32_t end_offset) {
-    bool result = false;
-    uint32_t intersection_min = std::max(static_cast<uint32_t>(start), static_cast<uint32_t>(end));
-    uint32_t intersection_max = std::min(static_cast<uint32_t>(start) + start_offset, static_cast<uint32_t>(end) + end_offset);
+// Returns true if [x, x + x_offset] and [y, y + y_offset] overlap
+static sparse_container::range<int64_t> GetRangeIntersection(int64_t x, uint64_t x_offset, int64_t y, uint64_t y_offset) {
```
Comment (artem-lunarg) [inline layers/core_checks/cc_copy_blit_resolve.cpp:30]: nit: name suggestion: `offset` name sometimes associates with signed values, `count` or `size` could be more intuitive for unsigned types and might be easier to validate arithmetic (some similarity with std containers that also use semi-closed intervals).

**Review by artem-lunarg — APPROVED** [review]
LGTM! Comment nits.

File: layers/core_checks/cc_copy_blit_resolve.cpp [hunk 4]
```diff
@@ -1982,12 +2035,18 @@ bool CoreChecks::ValidateCmdCopyImage(VkCommandBuffer commandBuffer, VkImage src
...
+        // so checking for memory overlaps is not possible.
         if (src_image_state->image() == dst_image_state->image()) {
             for (uint32_t j = 0; j < regionCount; j++) {
                 const LogObjectList objlist(commandBuffer, srcImage, dstImage);
-                if (RegionIntersects(&region, &pRegions[j], src_image_type, FormatIsMultiplane(src_format))) {
+                if (auto intersection = GetRegionIntersection(region, pRegions[j], src_image_type, FormatIsMultiplane(src_format));
+                    intersection.has_instersection) {
                     vuid = is_2 ? "VUID-VkCopyImageInfo2-pRegions-00124" : "VUID-vkCmdCopyImage-pRegions-00124";
-                    skip |= LogError(vuid, objlist, loc, "pRegion[%" PRIu32 "] src overlaps with pRegions[%" PRIu32 "].", i, j);
+                    skip |= LogError(vuid, objlist, loc,
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_copy_blit_resolve.cpp:2047]: > pRegion[0] src overlaps with pRegions[0]. 

I guess I don't understand this message, was it suppose to say

> pRegion[0] src overlaps with pRegions[0] dst. 

?

File: layers/core_checks/cc_copy_blit_resolve.cpp [hunk 5]
```diff
@@ -18,49 +18,95 @@
...
+struct ImageRegionIntersection {
+    VkImageSubresourceLayers subresource = {};
+    VkOffset3D offset = {0, 0, 0};
+    VkExtent3D extent = {1, 1, 1};
+    bool has_instersection = false;
+    std::string String() const noexcept {
+        std::stringstream ss;
+        ss << "{ subresource { aspectMask: " << string_VkImageAspectFlags(subresource.aspectMask)
+           << ", mipLevel: " << subresource.mipLevel << ", baseArrayLayer: " << subresource.baseArrayLayer
+           << ", layerCount: " << subresource.layerCount << " }, offset {" << offset.x << ", " << offset.y << ", " << offset.z
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_copy_blit_resolve.cpp:52]: > Overlap info: { subresource { aspectMask: VK_IMAGE_ASPECT_COLOR_BIT, mipLevel: 0, baseArrayLayer: 0, layerCount: 1 }, offset {0, 0, 0}, extent {64, 64, 1} }

I guess not sure if my src Region is

> offset {4, 4, 0}, extent {4, 4, 1} 

and my dst region is

 > offset {6, 6, 0}, extent {4, 1, 1} 
 
 what would this print out?

File: layers/core_checks/cc_copy_blit_resolve.cpp (hunk 4 above)
Comment (arno-lunarg) [reply layers/core_checks/cc_copy_blit_resolve.cpp:2047]: Oh yes wording is busted, do you think this is better:

vkCmdCopyImage(): pRegion[0] copy source overlaps with pRegions[0] copy destination. Overlap info, with respect to image (...)

**Review by spencer-lunarg — APPROVED** [review]

**Pull request merged**